javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll

```diff
@@ -13,106 +13,39 @@ module NoSQL {
 }
 
 /**
- * Gets a reference to an object where the "$where" property has been assigned to`rhs`.
+ * Gets a value that has been assigned to the "$where" property of an object that flows to `queryArg`.
  */
-DataFlow::SourceNode getADollarWherePropertyValueSource(DataFlow::TypeTracker t, DataFlow::Node rhs) {
-  t.start() and
-  rhs = result.getAPropertyWrite("$where").getRhs()
-  or
-  exists(DataFlow::TypeTracker t2 |
-    result = getADollarWherePropertyValueSource(t2, rhs).track(t2, t)
-  )
-}
-
-/**
- * Gets the value of a `$where` property of an object that flows to `n`.
- */
-private DataFlow::Node getADollarWherePropertyValue(DataFlow::Node n) {
-  getADollarWherePropertyValueSource(DataFlow::TypeTracker::end(), result).flowsTo(n)
+private DataFlow::Node getADollarWhereProperty(API::Node queryArg) {
+  result = queryArg.getMember("$where").getARhs()
 }
 
 /**
  * Provides classes modeling the MongoDB library.
  */
 private module MongoDB {
-  /**
-   * Gets an import of MongoDB.
-   */
-  DataFlow::ModuleImportNode mongodb() { result.getPath() = "mongodb" }
-
   /**
    * Gets an access to `mongodb.MongoClient`.
    */
-  private DataFlow::SourceNode getAMongoClient(DataFlow::TypeTracker t) {
-    t.start() and
-    (
-      result = mongodb().getAPropertyRead("MongoClient")
-      or
-      exists(DataFlow::ParameterNode p |
-        p = result and
-        p = getAMongoDbCallback().getParameter(1) and
-        not p.getName().toLowerCase() = "db" // mongodb v2 provides a `Db` here
-      )
-    )
-    or
-    exists(DataFlow::TypeTracker t2 | result = getAMongoClient(t2).track(t2, t))
-  }
-
-  /**
-   * Gets an access to `mongodb.MongoClient`.
-   */
-  DataFlow::SourceNode getAMongoClient() { result = getAMongoClient(DataFlow::TypeTracker::end()) }
-
-  /** Gets a data flow node that leads to a `connect` callback. */
-  private DataFlow::SourceNode getAMongoDbCallback(DataFlow::TypeBackTracker t) {
-    t.start() and
-    result = getAMongoClient().getAMemberCall("connect").getLastArgument().getALocalSource()
+  private API::Node getAMongoClient() {
+    result = API::moduleImport("mongodb").getMember("MongoClient")
     or
-    exists(DataFlow::TypeBackTracker t2 | result = getAMongoDbCallback(t2).backtrack(t2, t))
+    // slightly imprecise, is not supposed to have a result if the parameter name is "db" (that would be a mongodb v2 `Db`).
+    result = getAMongoDbCallback().getParameter(1)
```

Comment:
Can't you use the newly introduced `getAnImmediateUse` to rule this out? The immediate use would be the `DataFlow::Parameter` itself, so you can check its name.